lib/ace/edit_session/bracket_match.js

```diff
@@ -37,18 +37,115 @@ var Range = require("../range").Range;
 
 function BracketMatch() {
 
+
+    this.findSurroundingBrackets = function(position) {
+        if ((position.row == 0) && (position.column == 0)) return null;
+
+        var positionLeft = {row: position.row, column: position.column},
+            positionRight = {row: position.row, column: position.column};
+
+        var lineForLeftShift = this.getLine(positionLeft.row),
+            lineForRightShift = this.getLine(position.row);
+
+
+        var charBeforeCursor,
+            matchLeft,
+            charAfterCursor,
+            matchRight;
+        
+        var depthLeft = 0,
+            depthRight = 0;
+
+        var regExpOpenBrackets = /([\(\[\{])/,
+            regExpCloseBrackets = /([\)\]\}])/;
+
+        var iterator = new TokenIterator(this, positionLeft.row, positionLeft.column);
+        var tokens = iterator.getTokens();
+    
+
+        //Search in both directions until we find proper opening or closing bracket, afterwards we just find matching bracket in other direction.
+        while (!matchLeft && !matchRight) {
+
+
+            charBeforeCursor = lineForLeftShift.charAt(positionLeft.column);
+            charAfterCursor = lineForRightShift.charAt(positionRight.column);
+
+
+
+            if (charBeforeCursor.match(regExpCloseBrackets)) {
+                depthLeft++;
+            } else { 
+                matchLeft = charBeforeCursor.match(regExpOpenBrackets);
+                if ((matchLeft) && (depthLeft !== 0)) {
+                    matchLeft = null;
+                    depthLeft--;
+                }
+            }
+
+           
+
+            if (charAfterCursor.match(regExpOpenBrackets)) {
+                depthRight++;
+            } else {
+                matchRight = charAfterCursor.match(regExpCloseBrackets);
+                if ((matchRight) && (depthRight !== 0)) {
+                    matchRight = null;
+                    depthRight--;
+                }   
+            }
+
+            if (!charBeforeCursor) {
+                if (tokens[0]) {
+                    if (tokens[0].type === "keyword") {
+                        return null;
+                    } else {
+                        positionLeft.row--;
+                        lineForLeftShift = this.getLine(positionLeft.row);
+                        positionLeft.column = lineForLeftShift.length;
+                        iterator = new TokenIterator(this,positionLeft.row,positionLeft.column);
+                        tokens = iterator.getTokens();
+                    }
+                } else {
+                    return null;
+                }
+
+            }
+
+            if (!charAfterCursor) {
+                positionRight.row++;
+                lineForRightShift = this.getLine(positionRight.row);
+                positionRight.column = 0;
+            }
+
+            positionLeft.column--;
+            positionRight.column++;
+
+        }
+
+        if (matchLeft) {
+            positionLeft.column += 2;
```

Comment:
well, I am sure it is needed, but it used to be `-= length - 1`, now it is `+=2` which looks quite different. Just curious what is actually right. Care to write a comment explaining where positionLeft.column points to at this moment and why do we add and subtract 2 and 1?